addons/l10n_es_edi_verifactu/models/verifactu_document.py

```diff
@@ -0,0 +1,1139 @@
+            response_info = record_info.get(record_key, None)
+            if response_info is None:
+                response_info = {
+                    'errors': [_("We could not find any information about the record in the linked batch document.")],
+                }
+        else:
+            # I.e. in case of soapfault and access denied there is no `record_info`.
+            # So we just return the global 'state' / 'errors'.
+            response_info = {
+                'state': info['state'],
+                'errors': info['errors'],
+            }
+
+        # Add some information from the batch level in any case.
+        response_info.update({
+            'waiting_time_seconds': info.get('waiting_time_seconds', False),
+            'response_csv': info.get('response_csv', False),
+        })
+
+        return response_info
+
+    def _get_record_key(self):
+        self.ensure_one()
+        record_identifier = self.record_identifier
+        return str((record_identifier['IDEmisorFactura'], record_identifier['NumSerieFactura']))
+
+    def _post_send_hook(self, info):
+        # This function should not raise since it may be called "in the middle" of the sending process
+        for document in self:
+            invoice = document.move_id
+            if invoice.l10n_es_edi_verifactu_state == 'cancelled' and invoice.state != 'cancel':
+                with contextlib.suppress(UserError):
+                    invoice.button_cancel()
+
+    @api.ondelete(at_uninstall=False)
+    def _never_unlink_chained_documents(self):
+        for document in self:
+            if document.chain_index:
+                raise UserError(_("You cannot delete Veri*Factu Documents that are part of the chain of all Veri*Factu Documents."))
+
+    @api.model
+    def _format_date_fecha_type(self, date):
+        # Format as 'fecha' type from xsd
+        return date.strftime('%d-%m-%Y')
+
+    @api.model
+    def _round_format_number_2(self, number):
+        # Round and format as number with 2 precision digits
+        if number is None:
+            return None
+        rounded = float_round(number, precision_digits=2)
+        return float_repr(rounded, precision_digits=2)
+
+    # We do not check / fix the number of digits in front of the decimal separator
+    _format_number_ImporteSgn12_2 = _round_format_number_2
+    _format_number_Tipo2_2 = _round_format_number_2
+
+    @api.model
+    def _render_vals(self, vals, previous_record_identifier=None):
+        cancellation = vals['cancellation']
+        company = vals['company']
+        record_type = 'RegistroAnulacion' if cancellation else 'RegistroAlta'
+        render_vals = {
+            'company': company,
+            'record_type': record_type,
+            'record': vals['record'],
+            'cancellation': cancellation,
+            'errors': [],
+            'vals': vals,
+            'previous_record_identifier': previous_record_identifier,
```

Comment:
No need to have an intermediary method here.  You can immediately put everything immediately in the final form.